File: algo_2021/w5.-traverse-divide-and-conquer/problems/480.Binary_Tree_Paths/480_binary_tree_paths__travesal.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */

public class Solution {
    /**
     * @param root: the root of the binary tree
     * @return: all root-to-leaf paths
     */
    public List<String> binaryTreePaths(TreeNode root) {
        // write your code here
        List<String> res = new ArrayList<String>();
        
        if (root == null) {
            return res;
        }
        String path = String.valueOf(root.val);
        return findPath(root, path, res);
    }
    
    private List<String> findPath(TreeNode node, String path, List<String> res) {
        if (node == null) {
            return res;
        }
        if(node.left == null && node.right == null) {
            res.add(path);
            return res;
        }

        if (node.left != null) {
            // this way that every time called function would create new apth string
            // total times would be: 1 + 2 + ...+ n -> space complexity: O(n^2) 
            findPath(node.left, path + "->" + node.left.val, res);
        }
        if (node.right != null) {
            findPath(node.right, path + "->" + node.right.val, res);
        }
        return res;
    }
}

File: lintcode_algo_2021/s5.-traverse-d-and-c/480.Binary_Tree_Paths/480_binary_tree_paths__travesal.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: the root of the binary tree
    @return: all root-to-leaf paths
    """
    # travesal (DFS)
    def binaryTreePaths(self, root):
        # write your code here
        # result
        res = []
        if not root:
            res
        # put root node inside path
        path = [root]
        self.find_path(root, path, res)
        return res
    
    def find_path(self, node, path, res):
        if not node:
            return
        if not node.left and not node.right:
            res.append('->'.join([str(n.val) for n in path]))
            return res
        
        path.append(node.left)
        self.find_path(node.left, path, res)
        path.pop()

        path.append(node.right)
        self.find_path(node.right, path, res)
        path.pop()